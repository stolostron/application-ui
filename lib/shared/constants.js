/*******************************************************************************
 * Licensed Materials - Property of IBM
 * (c) Copyright IBM Corporation 2017, 2019. All Rights Reserved.
 * Copyright (c) 2021 Red Hat, Inc.
 *
 * US Government Users Restricted Rights - Use, duplication or disclosure
 * restricted by GSA ADP Schedule Contract with IBM Corp.
 *******************************************************************************/
'use strict'

module.exports.ACM_ACCESS_COOKIE = 'acm-access-token-cookie'
module.exports.ACM_AUTH_COOKIE = 'acm-acs-auth-cookie'
module.exports.ACM_USER_COOKIE = 'acm-user-cookie'
module.exports.ACM_REFRESH_INTERVAL_COOKIE = 'acm-refresh-interval-cookie'

module.exports.MCM_DIAGRAM_FILTER_COOKIE = 'mcm-diagram-filter-cookie'
module.exports.MCM_DESIGN_SPLITTER_OPEN_COOKIE =
  'mcm-design-splitter-open-cookie'
module.exports.MCM_DESIGN_SPLITTER_SIZE_COOKIE =
  'mcm-design-splitter-size-cookie'
module.exports.OVERVIEW_REFRESH_INTERVAL_COOKIE =
  'mcm-overview-interval-refresh-cookie'
module.exports.DIAGRAM_REFRESH_INTERVAL_COOKIE =
  'mcm-diagram-interval-refresh-cookie'
module.exports.REFRESH_TIMES = [15, 30, 60, 5 * 60, 30 * 60, 0]

module.exports.OVERVIEW_QUERY_COOKIE = 'mcm-overview-query-cookie'
module.exports.OVERVIEW_STATE_COOKIE = 'mcm-overview-state-cookie'
module.exports.DIAGRAM_QUERY_COOKIE = 'mcm-diagram-query-cookie'

module.exports.MAX_CHART_DATA_SIZE = 30

module.exports.LOCAL_HUB_NAME = 'local-cluster'

module.exports.RESOURCE_TYPES = {
  QUERY_APPLICATIONS: {
    name: 'QueryApplications',
    list: 'QueryApplicationList'
  },
  QUERY_SUBSCRIPTIONS: {
    name: 'QuerySubscriptions',
    list: 'QuerySubscriptionList'
  },
  QUERY_PLACEMENTRULES: {
    name: 'QueryPlacementRules',
    list: 'QueryPlacementRuleList'
  },
  QUERY_CHANNELS: {
    name: 'QueryChannels',
    list: 'QueryChannelList'
  },
  HCM_APPLICATIONS: { name: 'HCMApplication', list: 'HCMApplicationList' },
  HCM_CHANNELS: { name: 'HCMChannel', list: 'HCMChannelList' },
  HCM_SUBSCRIPTIONS: { name: 'HCMSubscription', list: 'HCMSubscriptionList' },
  HCM_PLACEMENT_RULES: {
    name: 'HCMPlacementRule',
    list: 'HCMPlacementRuleList'
  },
  HCM_FILTER_LIST: { name: 'HCMFilterList', list: 'HCMFilterList' },
  HCM_TOPOLOGY: { name: 'HCMTopology', list: 'HCMTopology' },
  USER_INFO: { name: 'userInfo', list: 'userInfoList' }
}

module.exports.MCM_CLUSTERS_SERVICES_ACTIONS = {}

module.exports.DOC_PATH_BASE =
  'https://access.redhat.com/documentation/en-us/red_hat_advanced_cluster_management_for_kubernetes/2.2/html/manage_applications/managing-applications'

module.exports.DOC_LINKS = {
  HOME: `${module.exports.DOC_PATH_BASE}`,
  TERMINOLOGY: `${
    module.exports.DOC_PATH_BASE
  }#application-model-and-definitions`,
  APPLICATIONS: `${
    module.exports.DOC_PATH_BASE
  }#managing-application-resources`,
  SUBSCRIPTIONS: `${
    module.exports.DOC_PATH_BASE
  }#creating-and-managing-subscriptions`,
  PLACEMENT_RULES: `${
    module.exports.DOC_PATH_BASE
  }#creating-and-managing-placement-rules`,
  CHANNELS: `${module.exports.DOC_PATH_BASE}#creating-and-managing-channels`
}
