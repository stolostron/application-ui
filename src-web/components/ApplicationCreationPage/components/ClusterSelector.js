/*******************************************************************************
 * Licensed Materials - Property of IBM
 * Copyright (c) 2020 Red Hat, Inc. All Rights Reserved.
 *
 * Note to U.S. Government Users Restricted Rights:
 * Use, duplication or disclosure restricted by GSA ADP Schedule
 * Contract with IBM Corp.
 *******************************************************************************/
'use strict'

import React from 'react'
import PropTypes from 'prop-types'
import resources from '../../../../lib/shared/resources'
import {
  Accordion,
  AccordionItem,
  Checkbox,
  Icon,
  TextInput
} from 'carbon-components-react'
import { getSourcePath, removeVs } from '../../TemplateEditor/utils/utils'
import Tooltip from '../../TemplateEditor/components/Tooltip'
import msgs from '../../../../nls/platform.properties'
import _ from 'lodash'

resources(() => {
  require('./style.scss')
})

export class ClusterSelector extends React.Component {
  static propTypes = {
    control: PropTypes.object,
    controlId: PropTypes.string,
    handleChange: PropTypes.func,
    locale: PropTypes.string
  };

  constructor(props) {
    super(props)
    this.state = {}
    if (_.isEmpty(this.props.control.active)) {
      if (
        !this.props.control.showData ||
        this.props.control.showData.length === 0
      ) {
        this.props.control.active = {
          mode: true,
          clusterLabelsList: [
            { id: 0, labelName: '', labelValue: '', validValue: false }
          ],
          clusterLabelsListID: 1
        }
      } else {
        //display existing placement rule
        this.props.control.active = {
          mode: true,
          clusterLabelsList: this.props.control.showData,
          clusterLabelsListID: this.props.control.showData.length
        }
      }
    }
    this.props.control.validation = this.validation.bind(this)
  }

  render() {
    if (!this.props.control.active) {
      this.props.control.active = {
        mode: true,
        clusterLabelsList: [
          { id: 0, labelName: '', labelValue: '', validValue: false }
        ],
        clusterLabelsListID: 1
      }
    }
    const { controlId, locale, control } = this.props
    const { name, active, validation = {} } = control
    const modeSelected = active && active.mode === true
    const isReadOnly = _.get(this.props, 'control.showData', []).length > 0

    return (
      <React.Fragment>
        <div className="creation-view-controls-labels">
          <div>
            {name}
            {validation.required ? (
              <div className="creation-view-controls-required">*</div>
            ) : null}
            <Tooltip control={control} locale={locale} />
          </div>

          <div className="clusterSelector-container">
            <Checkbox
              className="clusterSelector-checkbox"
              name="clusterSelector-checkbox"
              checked={modeSelected}
              disabled={isReadOnly}
              id={`clusterSelector-checkbox-${controlId}`}
              labelText={msgs.get(
                'tooltip.creation.app.settings.clusterSelector',
                locale
              )}
              onClick={this.handleChange.bind(this)}
            />

            <Accordion>
              <AccordionItem open={modeSelected} title={''}>
                <div className="clusterSelector-labels-section">
                  <div className="labels-descr">
                    {msgs.get(
                      'creation.app.settings.selectorClusters.config',
                      locale
                    )}
                  </div>

                  <div
                    className="labels-section"
                    id={`clusterSelector-labels-section-${controlId}`}
                  >
                    {this.renderClusterLabels(control, isReadOnly, controlId)}
                    <div
                      className={`add-label-btn ${
                        isReadOnly ? 'btn-disabled' : ''
                      }`}
                      tabIndex="0"
                      role={'button'}
                      onClick={() => this.addLabelToList(control, !isReadOnly)}
                      onKeyPress={this.addLabelKeyPress.bind(this)}
                    >
                      <Icon
                        name="icon--add--glyph"
                        fill="#3d70b2"
                        description=""
                        className="add-label-btn-icon"
                      />
                      <div className="add-label-btn-text">
                        {msgs.get(
                          'creation.app.settings.selectorClusters.prop.add',
                          locale
                        )}
                      </div>
                    </div>
                  </div>
                </div>
              </AccordionItem>
            </Accordion>
          </div>
        </div>
      </React.Fragment>
    )
  }

  validation(exceptions) {
    const { control, locale, controlId } = this.props
    if (_.get(control, 'active.mode', false)) {
      if (Object.keys(control.active.clusterLabelsList).length === 0) {
        //no cluster labels set
        exceptions.push({
          row: 1,
          text: msgs.get('creation.missing.clusterSelector.value', locale),
          type: 'error',
          controlId: `clusterSelector-labels-section-${controlId}`
        })
      }

      control.active.clusterLabelsList.map(item => {
        const { id, labelName, labelValue } = item
        const invalidLabel = !labelName || labelName.length === 0
        const invalidValue = !labelValue || labelValue.length === 0

        // Add exception if no input for labels or values
        if (invalidLabel) {
          exceptions.push({
            row: 1,
            text: msgs.get('creation.missing.clusterSelector.label', locale),
            type: 'error',
            controlId: `labelName-${id}`
          })
        }
        if (invalidValue) {
          exceptions.push({
            row: 1,
            text: msgs.get('creation.missing.clusterSelector.value', locale),
            type: 'error',
            controlId: `labelName-${id}`
          })
        }
      })
    }
  }

  renderClusterLabels = (control, isReadOnly, controlId) => {
    if (!_.get(control, 'active.clusterLabelsList')) {
      return ''
    }
    return (
      control.active &&
      control.active.clusterLabelsList.map(item => {
        const { id, labelName, labelValue, validValue } = item
        const invalidLabel = !labelName || labelName.length === 0
        const exceptionLabel = msgs.get(
          'creation.missing.clusterSelector.label.field'
        )
        const invalidValue = !labelValue || labelValue.length === 0
        const exceptionValue = msgs.get(
          'creation.missing.clusterSelector.value.field'
        )

        if (validValue || id === 0) {
          return (
            <React.Fragment key={id}>
              <div className="matching-labels-container">
                <div className="matching-labels-input">
                  <TextInput
                    id={`labelName-${id}-${controlId}`}
                    invalid={invalidLabel}
                    invalidText={exceptionLabel}
                    name="labelName"
                    className="text-input"
                    labelText={
                      id === 0
                        ? `${msgs.get('clusterSelector.label.field.ui')}*`
                        : ''
                    }
                    value={labelName === '' ? '' : labelName}
                    placeholder={msgs.get(
                      'clusterSelector.label.placeholder.field'
                    )}
                    disabled={isReadOnly}
                    onChange={this.handleChange.bind(this)}
                  />
                </div>
                <div className="matching-labels-input">
                  <TextInput
                    invalid={invalidValue}
                    invalidText={exceptionValue}
                    id={`labelValue-${id}-${controlId}`}
                    name="labelValue"
                    className="text-input"
                    labelText={
                      id === 0
                        ? `${msgs.get('clusterSelector.value.field.ui')}*`
                        : ''
                    }
                    value={labelValue === '' ? '' : labelValue}
                    placeholder={msgs.get(
                      'clusterSelector.value.placeholder.field'
                    )}
                    disabled={isReadOnly}
                    onChange={this.handleChange.bind(this)}
                  />
                </div>

                {id !== 0 ? ( // Option to remove added labels
                  <div
                    id={id}
                    className="remove-label-btn"
                    tabIndex="0"
                    role={'button'}
                    onClick={() => this.removeLabelFromList(control, item)}
                    onKeyPress={this.removeLabelKeyPress.bind(this)}
                  >
                    <Icon
                      name="icon--close--glyph"
                      fill="#3d70b2"
                      className="remove-label-btn-icon"
                    />
                  </div>
                ) : (
                  ''
                )}
              </div>
            </React.Fragment>
          )
        }
        return ''
      })
    )
  };

  addLabelToList = (control, modeSelected) => {
    if (modeSelected) {
      // Create new "label" item
      control.active.clusterLabelsList.push({
        id: control.active.clusterLabelsListID,
        labelName: '',
        labelValue: '',
        validValue: true
      })
      control.active.clusterLabelsListID++

      // Update UI
      this.forceUpdate()
    }
  };

  addLabelKeyPress = e => {
    if (e.type === 'click' || e.key === 'Enter') {
      this.addLabelToList(this.props.control)
    }
  };

  removeLabelFromList = (control, item) => {
    // Removed labels are no longer valid
    control.active.clusterLabelsList[item.id].validValue = false

    // Update UI and yaml editor
    this.forceUpdate()
    this.handleChange({})
  };

  removeLabelKeyPress = e => {
    if (e.type === 'click' || e.key === 'Enter') {
      this.removeLabelFromList(this.props.control, { id: e.target.id })
    }
  };

  handleChange(event) {
    const { control, handleChange } = this.props
    let targetName = ''
    try {
      targetName = event.target.name
    } catch (e) {
      targetName = ''
    }

    if (targetName) {
      const { active } = control
      if (targetName === 'clusterSelector-checkbox') {
        active.mode = event.target.checked ? true : false
      } else {
        const { clusterLabelsList } = active
        const labelID = parseInt(event.target.id.split('-')[1], 0)
        if (targetName === 'labelName') {
          clusterLabelsList[labelID].labelName = event.target.value
        } else if (targetName === 'labelValue') {
          clusterLabelsList[labelID].labelValue = event.target.value
        }
        clusterLabelsList[labelID].validValue = true
      }
    }

    handleChange(control)
  }
}

export default ClusterSelector

export const summarize = (control, controlData, summary) => {
  const { clusterLabelsList } = control.active || {}
  if (
    clusterLabelsList &&
    _.get(control, 'type', '') !== 'hidden' &&
    _.get(control, 'active.mode')
  ) {
    clusterLabelsList.forEach(({ labelName, labelValue }) => {
      if (labelName && labelValue) {
        summary.push(`${labelName}=${labelValue}`)
      }
    })
  }
}

export const reverse = (control, templateObject) => {
  if (!control.active) {
    let matchLabels
    const local = _.get(
      templateObject,
      getSourcePath(
        'PlacementRule[0].spec.clusterSelector.matchLabels.local-cluster'
      )
    )
    if (!local) {
      matchLabels = _.get(
        templateObject,
        getSourcePath('PlacementRule[0].spec.clusterSelector.matchLabels')
      )
      if (!matchLabels) {
        matchLabels = _.get(
          templateObject,
          getSourcePath('PlacementRule[0].spec.clusterLabels.matchLabels')
        )
      }
    }
    if (matchLabels) {
      matchLabels = removeVs(matchLabels)
      if (matchLabels) {
        const clusterLabelsList = Object.entries(matchLabels).map(
          ([labelName, labelValue], id) => {
            return {
              id,
              labelName,
              labelValue,
              validValue: true
            }
          }
        )
        control.active = {
          mode: true,
          clusterLabelsList,
          clusterLabelsListID: clusterLabelsList.length
        }
      }
    } else {
      const clusterLabelsList = [
        { id: 0, labelName: '', labelValue: '', validValue: false }
      ]
      control.active = {
        mode: false,
        clusterLabelsList,
        clusterLabelsListID: 1
      }
    }
  }
}
