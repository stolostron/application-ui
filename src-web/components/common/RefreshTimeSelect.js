/*******************************************************************************
 * Licensed Materials - Property of IBM
 * (c) Copyright IBM Corporation 2019. All Rights Reserved.
 *
 * Note to U.S. Government Users Restricted Rights:
 * Use, duplication or disclosure restricted by GSA ADP Schedule
 * Contract with IBM Corp.
 *******************************************************************************/
// Copyright (c) 2020 Red Hat, Inc.
'use strict'

import React from 'react'
import PropTypes from 'prop-types'
import { DropdownV2, Loading } from 'carbon-components-react'
import '../../../graphics/diagramIcons.svg'
import '../../../scss/refresh-time-select.scss'
import msgs from '../../../nls/platform.properties'
import config from '../../../lib/shared/config'

// get the cookie name from constants

import { ACM_REFRESH_INTERVAL_COOKIE } from '../../../lib/shared/constants'

export default class RefreshTimeSelect extends React.Component {
  static propTypes = {
    isReloading: PropTypes.bool,
    locale: PropTypes.string,
    refetchIntervalUpdate: PropTypes.func,
    refreshValues: PropTypes.array
  };

  constructor(props) {
    super(props)
    const { refetchIntervalUpdate } = props
    const pollInterval = getPollInterval()

    this.state = {
      pollInterval,
      refetchIntervalUpdate
    }
    this.handleChange = this.handleChange.bind(this)

    refetchIntervalUpdate({ doRefetch: false, interval: pollInterval })
  }

  UNSAFE_componentWillMount() {
    const { refreshValues, locale } = this.props
    if (refreshValues && refreshValues.length > 0) {
      this.autoRefreshChoices = refreshValues.map(pollInterval => {
        let label
        if (pollInterval >= 60) {
          label = msgs.get(
            'refresh.interval.minutes',
            [pollInterval / 60],
            locale
          )
        } else if (pollInterval !== 0) {
          label = msgs.get('refresh.interval.seconds', [pollInterval], locale)
        } else {
          label = msgs.get('refresh.interval.never', locale)
        }
        pollInterval *= 1000
        return { label, pollInterval }
      })
    }
  }

  handleClick = () => {
    const { refetchIntervalUpdate } = this.state
    const pollInterval = getPollInterval()

    this.doRefetchAction(refetchIntervalUpdate, pollInterval)
  };

  handleKeyPress = e => {
    const { refetchIntervalUpdate } = this.state
    const pollInterval = getPollInterval()

    if (e.key === 'Enter') {
      this.doRefetchAction(refetchIntervalUpdate, pollInterval)
    }
  };

  // this call does the actual refetch logic
  doRefetchAction = (refetchIntervalUpdate, pollInterval) => {
    refetchIntervalUpdate({
      doRefetch: true,
      interval: pollInterval
    })

    // resets the "doRefetch" flag after 1 second
    setTimeout(() => {
      refetchIntervalUpdate({
        doRefetch: false,
        interval: pollInterval
      })
    }, 1000)
  };

  handleChange = e => {
    const { selectedItem: { pollInterval } } = e
    const { refetchIntervalUpdate } = this.state
    // save interval to cookie
    savePollInterval(pollInterval)
    // update state with the new poll interval and trigger refetch
    refetchIntervalUpdate({ doRefetch: false, interval: pollInterval })
  };

  UNSAFE_componentWillReceiveProps() {
    this.setState(() => {
      return { doRefetch: false, interval: getPollInterval() }
    })
  }

  render() {
    const { locale } = this.props
    const { pollInterval } = this.state
    if (
      pollInterval !== undefined &&
      this.autoRefreshChoices &&
      this.autoRefreshChoices.length > 0
    ) {
      const refresh = msgs.get('refresh', locale)
      const { isReloading } = this.props
      const label = msgs.get('refresh.choose', locale)
      const idx = Math.max(
        0,
        this.autoRefreshChoices.findIndex(({ pollInterval: pi }) => {
          return pollInterval === pi
        })
      )

      return (
        // conditional style to adjust position if there is a breadcrumb header
        <div className="refresh-time-selection">
          {isReloading ? (
            <div className="reloading-container">
              <Loading withOverlay={false} small />
            </div>
          ) : (
            <div
              id="refreshButton"
              className="button"
              tabIndex="0"
              role={'button'}
              title={refresh}
              aria-label={refresh}
              onClick={this.handleClick}
              onKeyPress={this.handleKeyPress}
            >
              <svg className="button-icon">
                <use href={'#diagramIcons_autoRefresh'} />
              </svg>
            </div>
          )}
          <DropdownV2
            id="refreshDropdown"
            className="selection"
            label={label}
            ariaLabel={label}
            onChange={this.handleChange}
            inline={true}
            initialSelectedItem={this.autoRefreshChoices[idx].label}
            items={this.autoRefreshChoices}
          />
        </div>
      )
    }
    return null
  }
}

export const getPollInterval = () => {
  let pollInterval = config['featureFlags:liveUpdatesPollInterval'] || 0
  const savedInterval = localStorage.getItem(ACM_REFRESH_INTERVAL_COOKIE)
  if (savedInterval) {
    try {
      const saved = JSON.parse(savedInterval)
      if (saved.pollInterval !== undefined) {
        pollInterval = saved.pollInterval
      }
    } catch (e) {
      //
    }
  } else {
    savePollInterval(pollInterval)
  }
  return pollInterval
}

export const savePollInterval = pollInterval => {
  localStorage.setItem(
    ACM_REFRESH_INTERVAL_COOKIE,
    JSON.stringify({ pollInterval })
  )
}
