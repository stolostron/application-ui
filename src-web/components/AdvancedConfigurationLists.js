/*******************************************************************************
 * Licensed Materials - Property of IBM
 * (c) Copyright IBM Corporation 2018, 2019. All Rights Reserved.
 * Copyright (c) 2020 Red Hat, Inc.
 * US Government Users Restricted Rights - Use, duplication or disclosure
 * restricted by GSA ADP Schedule Contract with IBM Corp.
 *******************************************************************************/
'use strict'

import React from 'react'
import ResourceList from './common/ResourceList'
import ResourceTableModule from './common/ResourceTableModuleFromProps'
import { RESOURCE_TYPES } from '../../lib/shared/constants'
import getResourceDefinitions from '../definitions'
import { makeGetVisibleTableItemsSelector } from '../reducers/common'
import QuerySwitcher from './common/QuerySwitcher'

const AdvancedConfigurationLists = (props) => {

  const resourceType = RESOURCE_TYPES.QUERY_APPLICATIONS
  const staticResourceData = getResourceDefinitions(resourceType)
  const getVisibleResources = makeGetVisibleTableItemsSelector(resourceType)

  return (
    <ResourceList
      {...props}
      resourceType={resourceType}
      staticResourceData={staticResourceData}
      getVisibleResources={getVisibleResources}
      modules={[<ResourceTableModule key="deployments" definitionsKey="deploymentKeys" />]}
    >
      <QuerySwitcher
        key='switcher'
        queryParam="resource"
        options={
              [{ id: 'subscription', contents: 'Subscriptions' },
                { id: 'placementrule', contents: 'Placement rules'},
                {id: 'channel', contents: 'Channels'}]}
        defaultOption="subscription"
      />
    </ResourceList>
  )}

export default AdvancedConfigurationLists
