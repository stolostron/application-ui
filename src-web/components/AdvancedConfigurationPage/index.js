/** *****************************************************************************
 * Licensed Materials - Property of IBM
 * (c) Copyright IBM Corporation 2017, 2019. All Rights Reserved.
 *
 * US Government Users Restricted Rights - Use, duplication or disclosure
 * restricted by GSA ADP Schedule Contract with IBM Corp.
 *******************************************************************************/
// Copyright (c) 2020 Red Hat, Inc.

import React from 'react'
import msgs from '../../../nls/platform.properties'
import { connect } from 'react-redux'
import resources from '../../../lib/shared/resources'
import {
  updateSecondaryHeader,
  clearSuccessFinished
} from '../../actions/common'

import ApplicationDeploymentHighlights from '../ApplicationDeploymentHighlights'
import { withRouter } from 'react-router-dom'
import AdvancedConfigurationLists from '../AdvancedConfigurationLists'

/* eslint-disable react/prop-types */

resources(() => {
  require('./style.scss')
})

const mapDispatchToProps = dispatch => {
  return {
    clearSuccessFinished: () => clearSuccessFinished(dispatch),
    updateSecondaryHeaderFn: (title, tabs, mainButton) =>
      dispatch(
        updateSecondaryHeader(title, tabs, null, null, null, null, mainButton)
      )
  }
}

class AdvancedConfigurationPage extends React.Component {
  componentDidMount() {
    const {
      secondaryHeaderProps,
      updateSecondaryHeaderFn,
      locale
    } = this.props

    updateSecondaryHeaderFn(
      msgs.get(secondaryHeaderProps.title, locale),
      secondaryHeaderProps.tabs,
      secondaryHeaderProps.mainButton
    )
  }

  componentWillUnmount() {
    this.mutateFinished()
  }

  mutateFinished() {
    this.props.clearSuccessFinished()
  }

  render() {
    const { secondaryHeaderProps } = this.props

    return (
      <React.Fragment>
        <ApplicationDeploymentHighlights />
        <AdvancedConfigurationLists
          secondaryHeaderProps={secondaryHeaderProps}
        />
      </React.Fragment>
    )
  }
}

export default withRouter(
  connect(null, mapDispatchToProps)(AdvancedConfigurationPage)
)
