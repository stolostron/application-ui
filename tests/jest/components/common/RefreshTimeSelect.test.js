// Copyright (c) 2020 Red Hat, Inc.

"use strict";

import React from "react";
import { mount, shallow } from "enzyme";
import configureMockStore from "redux-mock-store";
import { REFRESH_TIMES } from "../../../../lib/shared/constants";
import { reduxStoreAppPipeline } from "../../components/TestingData";
import RefreshTimeSelect from "../../../../src-web/components/common/RefreshTimeSelect";
import renderer from "react-test-renderer";

const mockStore = configureMockStore();
const storeApp = mockStore(reduxStoreAppPipeline);

describe("RefreshTimeSelect component", () => {
  const refetchIntervalUpdate = jest.fn();

  it("renders as expected", () => {
    const component = renderer.create(
      <RefreshTimeSelect
        locale="en"
        isReloading={false}
        refetchIntervalUpdate={refetchIntervalUpdate}
        refreshValues={REFRESH_TIMES}
      />
    );
    expect(component.toJSON()).toMatchSnapshot();
  });

  it("renders as expected - reloading = true", () => {
    const component = renderer.create(
      <RefreshTimeSelect
        locale="en"
        isReloading={true}
        refetchIntervalUpdate={refetchIntervalUpdate}
        refreshValues={REFRESH_TIMES}
      />
    );
    expect(component.toJSON()).toMatchSnapshot();
  });

  jest.useFakeTimers();
  it("Click and Keydown actions", () => {
    const refreshTimeSelect = shallow(
      <RefreshTimeSelect
        locale="en"
        isReloading={false}
        refetchIntervalUpdate={refetchIntervalUpdate}
        refreshValues={REFRESH_TIMES}
      />
    );

    // manually trigger a props update
    refreshTimeSelect.setProps({ abc: 123 });

    refreshTimeSelect.find("#refreshButton").simulate("click");
    refreshTimeSelect.update();
    // wait 1 second for refresh button to be ready again
    jest.runOnlyPendingTimers();

    refreshTimeSelect
      .find("#refreshButton")
      .simulate("keyPress", { key: "Enter" });
    refreshTimeSelect.update();
    // wait 1 second for refresh button to be ready again
    jest.runOnlyPendingTimers();

    // not enter key
    refreshTimeSelect.find("#refreshButton").simulate("keyPress", { key: "x" });
    refreshTimeSelect.update();

    refreshTimeSelect.find("DropdownV2").simulate("change", {
      selectedItem: { label: "Refresh every 5m", pollInterval: 300000 }
    });
    refreshTimeSelect.update();

    refreshTimeSelect.find("#refreshButton").simulate("click");
    refreshTimeSelect.update();

    expect(setTimeout).toHaveBeenCalledTimes(4);
  });

  it("renders as expected - empty refresh values", () => {
    const component = renderer.create(
      <RefreshTimeSelect
        locale="en"
        isReloading={true}
        refetchIntervalUpdate={refetchIntervalUpdate}
        refreshValues={null}
      />
    );
    expect(component.toJSON()).toMatchSnapshot();
  });
});
