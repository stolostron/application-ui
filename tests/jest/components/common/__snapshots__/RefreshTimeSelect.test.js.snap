// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RefreshTimeSelect component renders as expected - empty refresh values 1`] = `null`;

exports[`RefreshTimeSelect component renders as expected - reloading = true 1`] = `
<div
  className="refresh-time-selection"
>
  <div
    className="reloading-container"
  >
    <div
      aria-live="assertive"
      className="bx--loading bx--loading--small"
    >
      <svg
        className="bx--loading__svg"
        viewBox="-75 -75 150 150"
      >
        <circle
          cx="0"
          cy="0"
          r="37.5"
        />
      </svg>
    </div>
  </div>
  <div
    aria-label="Choose refresh"
    className="bx--dropdown selection bx--list-box"
    onClick={[Function]}
    onKeyDown={[Function]}
    role="listbox"
    tabIndex="0"
  >
    <div
      aria-expanded={false}
      aria-haspopup={true}
      aria-label="open menu"
      className="bx--list-box__field"
      data-toggle={true}
      disabled={false}
      onBlur={[Function]}
      onClick={[Function]}
      onKeyDown={[Function]}
      role="button"
      tabIndex="0"
      type="button"
    >
      <span
        className="bx--list-box__label"
        htmlFor="refreshDropdown-input"
      >
        Disable refresh
      </span>
      <div
        className="bx--list-box__menu-icon"
      >
        <svg
          alt="Open menu"
          aria-label="Open menu"
          fillRule="evenodd"
          height="5"
          role="img"
          viewBox="0 0 10 5"
          width="10"
        >
          <title>
            Open menu
          </title>
          <path
            d="M0 0l5 4.998L10 0z"
          />
        </svg>
      </div>
    </div>
  </div>
</div>
`;

exports[`RefreshTimeSelect component renders as expected 1`] = `
<div
  className="refresh-time-selection"
>
  <div
    aria-label="Refresh"
    className="button"
    id="refreshButton"
    onClick={[Function]}
    onKeyPress={[Function]}
    role="button"
    tabIndex="0"
    title="Refresh"
  >
    <svg
      className="button-icon"
    >
      <use
        href="#diagramIcons_autoRefresh"
      />
    </svg>
  </div>
  <div
    aria-label="Choose refresh"
    className="bx--dropdown selection bx--list-box"
    onClick={[Function]}
    onKeyDown={[Function]}
    role="listbox"
    tabIndex="0"
  >
    <div
      aria-expanded={false}
      aria-haspopup={true}
      aria-label="open menu"
      className="bx--list-box__field"
      data-toggle={true}
      disabled={false}
      onBlur={[Function]}
      onClick={[Function]}
      onKeyDown={[Function]}
      role="button"
      tabIndex="0"
      type="button"
    >
      <span
        className="bx--list-box__label"
        htmlFor="refreshDropdown-input"
      >
        Disable refresh
      </span>
      <div
        className="bx--list-box__menu-icon"
      >
        <svg
          alt="Open menu"
          aria-label="Open menu"
          fillRule="evenodd"
          height="5"
          role="img"
          viewBox="0 0 10 5"
          width="10"
        >
          <title>
            Open menu
          </title>
          <path
            d="M0 0l5 4.998L10 0z"
          />
        </svg>
      </div>
    </div>
  </div>
</div>
`;
